CMakeLists.txt

```diff
@@ -132,34 +130,82 @@ endif()
 
 option(INSTALL_PRIVATE_HEADERS "installs private headers in the same location as the public ones" OFF)
 
-find_package(BlocksRuntime QUIET)
-if(NOT BlocksRuntime_FOUND)
+if(NOT CMAKE_SYSTEM_NAME STREQUAL Darwin)
   set(BlocksRuntime_INCLUDE_DIR ${PROJECT_SOURCE_DIR}/src/BlocksRuntime)
 
-  add_library(BlocksRuntime
-              STATIC
-                ${PROJECT_SOURCE_DIR}/src/BlocksRuntime/data.c
-                ${PROJECT_SOURCE_DIR}/src/BlocksRuntime/runtime.c)
-  set_target_properties(BlocksRuntime
+  include(ExternalProject)
+  ExternalProject_Add(SharedBlocksRuntime
+                      SOURCE_DIR
+                        ${PROJECT_SOURCE_DIR}/src/BlocksRuntime
+                      CMAKE_ARGS
+                        -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+                        -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+                        -DCMAKE_INSTALL_PREFIX=<INSTALL_DIR>
+                        -DCMAKE_MAKE_PROGRAM=${CMAKE_MAKE_PROGRAM}
+                        -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TOOLCHAIN_FILE}
+                        -DBUILD_SHARED_LIBS=YES
+                      BUILD_BYPRODUCTS
+                        <BINARY_DIR>/${CMAKE_SHARED_LIBRARY_PREFIX}BlocksRuntime${CMAKE_SHARED_LIBRARY_SUFFIX}
+                        <BINARY_DIR>/${CMAKE_IMPORT_LIBRARY_PREFIX}BlocksRuntime${CMAKE_IMPORT_LIBRARY_SUFFIX}
+                      BUILD_ALWAYS
+                        1)
```

Comment:
My experience using `ExternalProject` hasn't been great compared to creating library targets that are directly addressable in the build system and I strongly suggest we don't use it in dispatch.

1. We need to propagate all possible interesting CMake variables from the outer build system to the inner ExternalProject build system via `CMAKE_ARGS`. For example here you don't propagate `CMAKE_C_FLAGS`, `CMAKE_AR` or `CMAKE_LIBRARY_PATH` which could cause the downstream libraries to be built differently. There are too many influential variables to enumerate here.
1. ExternalProjects don't get good dependency tracking compared to targets that cmake understands natively. You have to declare BUILD_BYPRODUCTS manually and it's easy to under or over-depend. ExternalProjects also don't have great IDE integration and they appear opaque.
1. Setting up a library target as `IMPORTED` prevents dependency tracking and prevents it being exported from the build system (via `install(EXPORT)` for example).
1. Ninja is not designed to be invoked recursively - you can get into a situation where the downstream ninja is performing tasks equal to the number of CPUs, so you can have (N^2-1) tasks running in parallel (where `N` is the number of cores your build machine has). Console output is also broken.

Having said all that, have you considered making `BlocksRuntime` and `BlocksRuntime_static` targets using `add_library` instead?

Reply:
Well, we already are using ExternalProject for libdispatch in swift, and it has been working quite well :-).  That said, there is interest in moving to the LLVM wrapper.  Doing so will automatically propagate those other variables implicitly.

The dependency tracking issue that you portray is not entirely correct.  There is tracking of the dependencies in the sense that the external project's byproducts are linked into the imported library target.  So if the sub-library changes, CMake/ninja realize that and rebuild the dependent targets.

Yes, I did consider that.  I suppose that if we want to just duplicate the rules, thats certainly an option.  I really am strongly opposed to creating the custom wrappers for both variants.  I suppose that we could always extract the BlocksRuntime rules into a separate CMakeList and include it in this one to make this easier to work with.